t/apicast-policy-rate-limit.t

```diff
@@ -0,0 +1,995 @@
+                      burst = 0,
+                      delay = 2
+                    }
+                  },
+                  redis_url = "redis://localhost:6379/1"
+                }
+              },
+              {
+                name = "apicast.policy.rate_limit",
+                configuration = {
+                  limiters = {
+                    {
+                      name = "connections",
+                      key = "test7",
+                      conn = 1,
+                      burst = 0,
+                      delay = 2
+                    }
+                  },
+                  redis_url = "redis://localhost:6379/1"
+                }
+              }
+            }
+          }
+        }
+      }
+    })
+  }
+  lua_shared_dict limiter 1m;
+
+--- config
+  include $TEST_NGINX_APICAST_CONFIG;
+
+  location /flush_redis {
+    content_by_lua_block {
+      local redis = require('resty.redis'):new()
+      redis:connect('127.0.0.1', 6379)
+      redis:select(1)
+      redis:del("test7")
+    }
+  }
+
+--- pipelined_requests eval
+["GET /flush_redis","GET /"]
+--- error_code eval
+[200, 429]
+--- no_error_log
+[error]
+
+=== TEST 8: Rejected (req).
+Return 503 code.
+--- http_config
+  include $TEST_NGINX_UPSTREAM_CONFIG;
+  lua_package_path "$TEST_NGINX_LUA_PATH";
+
+  init_by_lua_block {
+    require "resty.core"
+    ngx.shared.limiter:flush_all()
+    require('apicast.configuration_loader').mock({
+      services = {
+        {
+          id = 42,
+          proxy = {
+            policy_chain = {
+              {
+                name = "apicast.policy.rate_limit",
+                configuration = {
+                  limiters = {
+                    {
+                      name = "leaky_bucket",
+                      key = "test8",
+                      rate = 1,
+                      burst = 0
+                    }
+                  },
+                  redis_url = "redis://localhost:6379/1",
+                  status_code_rejected = 503
+                }
+              },
+              {
+                name = "apicast.policy.rate_limit",
+                configuration = {
+                  limiters = {
+                    {
+                      name = "leaky_bucket",
+                      key = "test8",
+                      rate = 1,
+                      burst = 0
+                    }
+                  },
+                  redis_url = "redis://localhost:6379/1",
+                  status_code_rejected = 503
+                }
+              }
+            }
+          }
+        }
+      }
+    })
+  }
+  lua_shared_dict limiter 1m;
+
+--- config
+  include $TEST_NGINX_APICAST_CONFIG;
+
+  location /flush_redis {
+    content_by_lua_block {
+      local redis = require('resty.redis'):new()
+      redis:connect('127.0.0.1', 6379)
+      redis:select(1)
+      redis:del("test8")
+    }
+  }
+
+--- pipelined_requests eval
+["GET /flush_redis","GET /"]
+--- error_code eval
+[200, 503]
```

Comment:
I don't think this is correct. `rate` is set to 1 in the config, so the first `GET /` should return 200 for sure. The second one would return 503 only if it was done in the same second, which is very likely in this test but might be a bit difficult to ensure that it's done that way.

Reply:
@davidor I define the policy twice here, so we can open 2 connections by "GET /", and the first "GET /" is a 503.

Reply:
I see now why you duplicated the policy in the config in several of these tests.

I find this way of testing the policy a bit counter-intuitive. To use this policy, users should include it just once and add there all the limiters they need. The problem I see with these tests is that the policy appears duplicated for testing convenience, but the configuration does not reflect a realistic scenario, because users will include the policy just once.

I think we should try to include the policy just once and change the tests so they work with that configuration. For example, in this particular test, we can make 3 requests. The first one to `/flush_redis`  (to make sure we start from a clean state), the second one to `/`(which should return 200), and the third one to `/` (which should return 503 because we defined a limit of 1 in the configuration).

What do you think?

Reply:
@davidor Your method works well for "leaky_bucket" and "fixed_window", but doesn't work well for "connections", I think. Regarding "connections", the second request returns 200 and then the number of connections is reset to zero, so the third request also returns 200.
However, I strongly agree with that the configuration does not reflect a realistic scenario in these tests, so I will change the tests of "leaky_bucket" and "fixed_window".

Reply:
This policy has to be reusable by multiple services at the same time. It will be loaded and active more than once in the same time.

We can't be defining shdicts per policy, so they all have to reuse one shdict. If some parts do not support that we can't support them. I think they should be supported because we control the rate limiting keys and can for example prefix them with the rate limiter name, so they are unique per rate limiter.

This is duplicated discussion of https://github.com/3scale/apicast/pull/648#discussion_r178231827.

Reply:
@mikz Let me confirm about your comment. Do you mean I should change the integration test as following?

before:
policy_chain = {{name = "apicast.policy.rate_limit", ...},{name = "apicast.policy.rate_limit", ...}}

after:
policy_chain = {{name = "apicast.policy.rate_limit", 
configuration = {limiters = {name = "connections", ...},{name = "connections", ...}}}}

Reply:
I'm not really saying that.

I'm saying there can be several instances of connection/leaky bucket/fixed window rate limiting. APIcast can run several services and in case of our Cloud Hosted deployment it runs several thousands of services and configurations - each of them could use one or more rate limiting policies - and still have to be using just one or two shared dictionaries.

No matter how many rate limiting policies you have they have to be able to run from ideally one shared dictionary. 

Because each policy can have unique rate limiting "key" ( https://github.com/openresty/lua-resty-limit-traffic/blob/master/lib/resty/limit/req.md#incoming , passed to .combine as table of keys https://github.com/openresty/lua-resty-limit-traffic/blob/master/lib/resty/limit/traffic.md#combine) they could operate on same shared dictionary, but just using different keys.

Now - defining those keys should be left to the user so they can implement for example cross service rate limiting. But if the same key being used between fixed window / leaky bucket / connection limiter causes trouble we can prefix those keys with limiter name. So user can still do cross service rate limiting but only with the same limiter (user setting key "foo" would use it as "fixed_window-foo" or "leaky_bucket-foo" or "connections-foo"). My understanding is those keys can't be shared between limiters anyway.

Reply:
@mikz Now I defined only one shdict called 'limiter' in nginx.conf.liquid. But I have to let the user to use multiple limiters with one shdict, right? The way I came up with is using FFI cdata as a value (for example [here](https://github.com/openresty/lua-resty-limit-traffic/blob/master/lib/resty/limit/req.lua#L111)). FFI cdata has the actual key and the value of limiter. Is my image matching with yours?

Reply:
@mikz I misunderstood how shdict works. Could you please review the latest commit?